### PR TITLE
safer use of api library + fixed panic 

### PR DESCRIPTION
**Pull request by artem-webdev**
# what do you think, maybe you shouldn't return nil, but rather close the channel?

#### in this case it will not be necessary to return the function before reading from the channel
```
go func(){

       chanRead, err := wac.Read(message.Info.RemoteJid, message.Info.Id)
        if err != nil {
            log.Println(err)   
       }

        // there will be no blocking even with this error ! 
        <-chanRead

}()

```

### in the current implementation with explicit nil
``` 
go func(){

       chanRead, err := wac.Read(message.Info.RemoteJid, message.Info.Id)
        if err != nil {
            log.Println(err)   
       }

        // nil channel leading to eternal blocking in case of programmer error ! 
        <-chanRead

}()

```

### CODE REVIEW

**Comment by artem-webdev** [comment]
I apologize for the many titles, I just didn't get enough sleep here is the latest version

**Comment by beshoo** [comment]
Again, how l see the panic on the current lib.
You are talking about "security" what is the security leak?

I can not see how can l approve this pl

**Comment by artem-webdev** [comment]
не закрывайте pl так быстро, вы можете увидеть изменения здесь https://github.com/Rhymen/go-whatsapp/pull/512/files , если вы не видите исправленных ошибок в коде, я могу !

**Comment by artem-webdev** [comment]
what exactly is not clear to you from the first comment with code examples?

**Comment by beshoo** [comment]
I closed the pl when it is NOT organized, l have no problem with the pl it self.
Thank you, but anyway, l stll can not see the ""security"" problem which you are referring to. Nor how panic may happens. Show us a panic report whuch you addressed with this PL.

Thanks

**Comment by beshoo** [comment]
My question is straightforward.
Where is the security lake? 
Whats is the panic you are addressing? 
What is the panic report? 

That is the title of the PL, if you are fine tuning the api then please tell the expected result. 

Regard

**Comment by artem-webdev** [comment]
if you are not good at code, then let others see it, it's still an open source project!

**Comment by beshoo** [comment]
I do appreciate your Pl, an i am good in golang, but that's is not the problem,
Your modification is appreciated. 
Any way l will let the community to decide. 
I can not approve this PL since it lake of details.

**Review by SchulteMK — CHANGES_REQUESTED** [review]
I agree that returning a closed channel is better than a nil channel.

Backround:
- Reading from a channel that is `nil ` (or not initialized) is not possible
  - if you try, you will wait forever
  - if you use select, that branch never gets selected 
- Reading from a closed channel will return instantly with the null-value of the channel type
  - it would never block

[Source](https://nanxiao.gitbooks.io/golang-101-hacks/content/posts/nil-channel-vs-closed-channel.html)

File: read.go [hunk 1]
```diff
@@ -83,7 +83,7 @@ func (wac *Conn) processReadData(msgType int, msg []byte) error {
 		// chan string to something like chan map[string]interface{}. The unmarshalling
 		// in several places, especially in session.go, would then be gone.
 		listener <- data[1]
-
+		close(listener)
 		wac.listener.Lock()
 		delete(wac.listener.m, data[0])
 		wac.listener.Unlock()
```
[the remark this answers is not in the document]
Comment (artem-webdev) [reply read.go]: thanks for your understanding, added your recommendation !

**Review by SchulteMK — APPROVED** [review]

**Pull request merged**